### ISSUE TITLE
asserts that return type is null | Generic<T> is reported as `null` and lose the generic

### ISSUE THREAD

**Issue by jdeniau** [microsoft/TypeScript#49588]
# Bug Report

### 🔎 Search Terms

- asserts generic null 
- asserts array null 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (3.7 on the playground as asserts did appear in 3.7 IIRC), and I reviewed the FAQ for entries about null asserts

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220616#code/C4TwDgpgBAKglgYwNYWFAvFA3gXwFCiSyIrAAycAzmpgHYCuANo1AD5QCCATlwIYgAealzi0A5m2LJUAPjx4AZvVoJgcAPa0ovSpQhdgAJQiNeazQElKAOXXBrTRgIBqMgBQA3AFxRnkhswAlD46egaUUB5QVFC2tA7MvABGjBAuMtj4SirmWqH6RiZmGrRWCYwA8lwU1BUKAMrAIuKePtx8gsKiEuzw0sAZ7AGMwdq6BRFRMe38Qk3dg1DDmYrKqiVQwBDUNcB1xqa5bmr9uz59pLuB2HhQmySouwDct1AA9G93XwB6APzyd3yBgOxUsNkcVV2dUazTExwe5CowECAPupyRLzuHy+dz+r2xEAAHpBVBAACabcDQGIAA2W7BmnXm4hkNLw+AIVKkpBBuQwS0cki64kkF1QqxyGyBhUOJTKEK4MO6rQFzCFzJ63NQo2lk2iEWFmuWuHk2XWmk22xloNo8P6B3OCN5JWuWFeJx5Jkx70+OLxr2lzrB5SqSpaHtQBxR7qdXvxvp+-yxnzJ6m2UAA7uouEgoApRNAM3BgAALdT0NA0xlsjmiLZcBS8BDQADiEFo+kQAhgGTddw8vEY9Ag5xeHLNfMDRVy8uYVTbHZECGhGpVC87CDmsNFCMWwx143CkX1UHXS63C38gpNnKIYuAZ8Q-Ppp-bG4vIt6u4l5q0W2oj4IHapCAY6-SAa6MbgW+S7etifr-AGh7WjO4JzlwgErrCwGoBB8ZTrKwZ2OUOEPjBiDRncEZkYuiBwQmuJJj6UBEiSWwUoQ1IRHSgrsIBH5iKy7JAA)

### 💻 Code

```ts
interface Generic<T> { value: T; }
type Ticket = {}
type TicketGeneric = null | Generic<string | Ticket>

function assertRelationIsNullOrGenericOfString(v: Generic<string | Ticket> | null): asserts v is Generic<string> | null {}

function testGeneric(ticketGeneric: TicketGeneric) {
  assertRelationIsNullOrGenericOfString(ticketGeneric)

  ticketGeneric;  // expected type is `null | Generic<string>` but gave `null`
}
```

### 🙁 Actual behavior

When an asserter return `null` or a Generic, it is reported as `null`.
It does this for every generic, even `Array<string>`. If the return type is not a generic, it does work fine (with `null | string` for exemple)

### 🙂 Expected behavior

It should be marked as `null | Generic<string>`

Thank you!

**Comment by MartinJohns** [comment]
I believe this is a duplicate of #31156. Assertion functions are essentially like type guards. Same issue is reported in #37807, which was named a duplicate.

**Comment by RyanCavanaugh** [comment]
This example doesn't have the subtype/assignability subtlety going on, but the general filter vs narrow behavior of UDTAs is intentional
